### PR TITLE
New Code Example: Pre-training BERT on NSP+MLM using Hugging Face Transformers

### PR DESCRIPTION
**Pull request by Sreyan88**
This code example shows how to pre-train BERT using Hugging Face transformers on NSP + MLM task. Additionally, it also shows best practices for Hugging Face Datasets.

### CODE REVIEW

**Comment by Sreyan88** [comment]
Hi @fchollet,

This is part of my GSoC 2022 contribution. Hope you can review it and comment!

Thank You so much!

**Comment by fchollet** [comment]
Thanks for the PR!

> This is part of my GSoC 2022 contribution

Cool, can you provide details? Who are you working with?

**Comment by Sreyan88** [comment]
Thank You for the reply. My mentor is @vasudevgupta7

File: examples/nlp/pretraining_BERT.py [hunk 1]
```diff
@@ -0,0 +1,481 @@
+"""
+Title: Pre-training BERT with 🤗 Transformers.
```
Comment (fchollet) [inline examples/nlp/pretraining_BERT.py:2]: Use "Hugging Face", do not include emojis in titles or descriptions.

File: examples/nlp/pretraining_BERT.py [hunk 2]
```diff
@@ -0,0 +1,481 @@
...
+allows the model to learn the context of a word based on all of its surroundings
+(left and right of the word).
+
+When training language models, there is a challenge of defining a prediction goal.
+Many models predict the next word in a sequence (e.g. The child came home from _ ),
+a directional approach which inherently limits context learning. To overcome this
+challenge, BERT uses two training strategies:
+
+### Masked Language Modeling (MLM)
+Before feeding word sequences into BERT, 15% of the words in each sequence are replaced
```
Comment (fchollet) [inline examples/nlp/pretraining_BERT.py:42]: Add line break above

File: examples/nlp/pretraining_BERT.py [hunk 3]
```diff
@@ -0,0 +1,481 @@
...
+with a `[MASK]` token. The model then attempts to predict the original value of the masked
+words, based on the context provided by the other, non-masked, words in the sequence.
+
+### Next Sentence Prediction (NSP)
+In the BERT training process, the model receives pairs of sentences as input and learns to
```
Comment (fchollet) [inline examples/nlp/pretraining_BERT.py:47]: Add line break above

File: examples/nlp/pretraining_BERT.py [hunk 4]
```diff
@@ -0,0 +1,481 @@
...
+
+import nltk
+import random
+
+import tensorflow as tf
+from tensorflow import keras
+
+nltk.download("punkt")
+# Only log error messages
+tf.compat.v1.logging.set_verbosity(tf.compat.v1.logging.ERROR)
```
Comment (fchollet) [inline examples/nlp/pretraining_BERT.py:88]: Do not use `compat.v1` for any reason. It is legacy and subject to future removal.

File: examples/nlp/pretraining_BERT.py [hunk 5]
```diff
@@ -0,0 +1,481 @@
...
+"""
+
+# We define the maximum number of tokens after tokenization that each training sample
+# will have
+max_num_tokens = BLOCK_SIZE - tokenizer.num_special_tokens_to_add(pair=True)
+
+
+def prepare_train_features(examples):
+
+    """
```
Comment (fchollet) [inline examples/nlp/pretraining_BERT.py:234]: Please use standard Keras formatting for docstrings

File: examples/nlp/pretraining_BERT.py [hunk 6]
```diff
@@ -0,0 +1,481 @@
...
+        # We *usually* want to fill up the entire sequence since we are padding
+        # to `block_size` anyways, so short sequences are generally wasted
+        # computation. However, we *sometimes*
+        # (i.e., short_seq_prob == 0.1 == 10% of the time) want to use shorter
+        # sequences to minimize the mismatch between pretraining and fine-tuning.
+        # The `target_seq_length` is just a rough target however, whereas
+        # `block_size` is a hard limit.
+        target_seq_length = max_num_tokens
+
+        if random.random() < SHORT_SEQ_PROB:
```
Comment (fchollet) [inline examples/nlp/pretraining_BERT.py:288]: Since you use randomness, please also seed your code via `keras.utils.set_random_seed`

File: examples/nlp/pretraining_BERT.py [hunk 7]
```diff
@@ -0,0 +1,481 @@
...
+
+optimizer = keras.optimizers.Adam(learning_rate=LEARNING_RATE)
+
+model.compile(optimizer=optimizer, run_eagerly=True)
+
+"""
+Finally all steps are done and now we can start training our model!
+"""
+
+model.fit(train, epochs=MAX_EPOCHS)
```
Comment (fchollet) [inline examples/nlp/pretraining_BERT.py:463]: Make sure to monitor validation data performance.

File: examples/nlp/pretraining_BERT.py [hunk 8]
```diff
@@ -0,0 +1,481 @@
...
+
+"""
+Now you can push this model to 🤗 Model Hub and also share it with with all your friends,
```
Comment (fchollet) [inline examples/nlp/pretraining_BERT.py:466]: Before, please add a section on what you can actually do with the model

File: examples/nlp/pretraining_BERT.py [hunk 9]
```diff
@@ -0,0 +1,481 @@
...
+model = TFBertForPreTraining(config)
+
+"""
+Now we will define our optimizer and compile the model. The loss calculation is handled
+internally and so we need not worry about that!
+"""
+
+optimizer = keras.optimizers.Adam(learning_rate=LEARNING_RATE)
+
+model.compile(optimizer=optimizer, run_eagerly=True)
```
Comment (fchollet) [inline examples/nlp/pretraining_BERT.py:457]: Why run eagerly? Typically this is only for debugging.

**Review by fchollet — COMMENTED** [review]
Thanks for the PR -- here are some initial comments.

In general there is rather little Keras functionality at play here. Could we modify the example so it works better as a Keras tutorial / showcase?

**Comment by Sreyan88** [comment]
Hi @fchollet ,

Thank You so much for the reply and initial comments! Really appreciate it and I will incorporate the changes as suggested.

> In general there is rather little Keras functionality at play here. Could we modify the example so it works better as a Keras tutorial / showcase?

I am actually a bit new to Keras/Tensorflow and was primarily following [this notebook](https://keras.io/examples/nlp/question_answering/) (the formatting and structure perspectives) for my tasks (my task is to create NLP examples for Keras using Hugging Face). Could you, if possible, suggest something that would help me satisfy the needs better for this?

Thank You again and hope you also find some time to review my other PRs ( #967 and #968 ).

File: examples/nlp/pretraining_BERT.py (hunk 9 above)
Comment (Sreyan88) [reply examples/nlp/pretraining_BERT.py]: Removed eager execution!

File: examples/nlp/pretraining_BERT.py (hunk 8 above)
Comment (Sreyan88) [reply examples/nlp/pretraining_BERT.py:492]: Added comment on how the pre-trained model can be used to fine-tune on any downstream task!

File: examples/nlp/pretraining_BERT.py (hunk 6 above)
Comment (Sreyan88) [reply examples/nlp/pretraining_BERT.py:298]: Added this!

File: examples/nlp/pretraining_BERT.py (hunk 5 above)
Comment (Sreyan88) [reply examples/nlp/pretraining_BERT.py]: Completed this!

File: examples/nlp/pretraining_BERT.py (hunk 1 above)
Comment (Sreyan88) [reply examples/nlp/pretraining_BERT.py]: Changed this!

File: examples/nlp/pretraining_BERT.py [hunk 10]
```diff
@@ -0,0 +1,513 @@
...
+from datasets import load_dataset
+
+dataset = load_dataset("wikitext", "wikitext-2-raw-v1")
+
+"""
+The dataset just has one column which is the raw text, and this is all we need for
+pre-training BERT!
+"""
+
+dataset
```
Comment (fchollet) [inline examples/nlp/pretraining_BERT.py:136]: The file should serve as a standalone Python script, so anything you want to display should use `print()`. Don't assume notebook behavior

File: examples/nlp/pretraining_BERT.py [hunk 11]
```diff
@@ -0,0 +1,513 @@
...
+- Create additional inputs for the model like `token_type_ids`, `attention_mask`, etc.
+"""
+
+# We define the maximum number of tokens after tokenization that each training sample
+# will have
+max_num_tokens = BLOCK_SIZE - tokenizer.num_special_tokens_to_add(pair=True)
+
+
+def prepare_train_features(examples):
+
```
Comment (fchollet) [inline examples/nlp/pretraining_BERT.py:237]: Remove line

File: examples/nlp/pretraining_BERT.py [hunk 12]
```diff
@@ -0,0 +1,513 @@
...
+    """Function to prepare features for NSP task
+
+    Attributes:
```
Comment (fchollet) [inline examples/nlp/pretraining_BERT.py:240]: Arguments (it's a function)

File: examples/nlp/pretraining_BERT.py [hunk 13]
```diff
@@ -0,0 +1,513 @@
...
+
+                    examples["input_ids"].append(padded["input_ids"])
+                    examples["token_type_ids"].append(padded["token_type_ids"])
+                    examples["attention_mask"].append(padded["attention_mask"])
+
+                    examples["next_sentence_label"].append(1 if is_random_next else 0)
+
+                    current_chunk = []
+                    current_length = 0
+
```
Comment (fchollet) [inline examples/nlp/pretraining_BERT.py:370]: Remove  blank line

**Review by fchollet — COMMENTED** [review]
Thank you for your patience -- I will proceed to review your other examples shortly as well.

Please address the comments -- before merging I will do a round of copyedits.

**Review by fchollet — COMMENTED** [review]
Thanks for the update! I have pushed some copyedits, please pull them first. Afterwards, you can add the generated files to the PR.

**Comment by Sreyan88** [comment]
Hi @fchollet ,

I think you can merge this now! I was also hoping if you could review my other PRs! Thank You again for your time!

**Comment by fchollet** [comment]
> cla/google Failing after 2s — ❌ Missing CLA from one or more contributors

Please sign the CLA!

Sorry for the delay in reviewing -- we need more folks than just me looking at these PRs.

**Comment by Sreyan88** [comment]
> > cla/google Failing after 2s — ❌ Missing CLA from one or more contributors
> 
> Please sign the CLA!
> 
> Sorry for the delay in reviewing -- we need more folks than just me looking at these PRs.

@fchollet, this PR messed up a bit with a push from my teammate's usernames due to shared machines. All ready to merge in #1044 . Thank You! I will close this after that is merged!

**Comment by fchollet** [comment]
The PR is merged, closing this now. Thanks!
